[quant][fx] Support some default ops in the native backend config

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74603
* #74601
* __->__ #74600

Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXGraphMatcher

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

jerryzh168 committed:
[quant][fx] Support some default ops in the native backend config

Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

andrewor14 commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -19,6 +29,25 @@ def get_native_backend_config_dict():
         "output_dtype": torch.quint8
     }
+
+    # input_activation: quint8, output_activation: quint8
```
nit: make these comments consistent across different dtype configs? Also I feel some of these comments are actually not needed, like this one doesn't give much additional information IMO

jerryzh168 replied:
sounds good, I can remove these

andrewor14 commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -19,6 +29,25 @@ def get_native_backend_config_dict():
+        # optional, output activation dtype
+        "output_dtype": torch.float16,
+    }
+    # END dtype configs
```
nit: I wonder if we should define these configs outside of `get_native_backend_config_dict` to make it easier to read, maybe in a future PR, otherwise this function will grow really large

andrewor14 replied:
Such that `get_native_backend_config_dict` will only look like this:
```
def get_native_backend_config_dict():
    """
    ...
    """
    return {
        # optional
        "name": "native",
        "configs": [
            linear_module_config,
            *default_op_configs,
            *default_op_fp16_configs,
             ...
        ],
    }
```
What do you think?

jerryzh168 replied:
yeah mean define everything outside? I think we may want to hide some of the implementation details, don't really want to expose everything in native.py, but we can expose the dtype configs, it might be useful for user as well

jerryzh168 replied:
I guess if we prefix these things with `_` it could work

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_numeric_suite_fx.py:
```diff
@@ -681,9 +680,9 @@ def _op_is_unmatchable(op):
                     _op_in_base_sets_of_related_ops(base_op),
                     f"{base_op} not in sets of related ops")
-            else:
+            elif not _op_in_base_sets_of_related_ops(base_op):
```
@vkuzo this is the change I made to the test, please take a look

jerryzh168 replied:
basically we don't explicitly check the quantize handler type in this branch, but just directly check if op is in the related ops set

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_numeric_suite_fx.py:
```diff
@@ -682,8 +678,14 @@ def _op_is_unmatchable(op):
             else:
-                raise AssertionError(
-                    f"handing for {qhandler_cls} not implemented")
+                # torch.sum does not have quantized equivalents
```
here is the change

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

jerryzh168 commented on test/quantization/fx/test_numeric_suite_fx.py:
```diff
@@ -682,8 +678,14 @@ def _op_is_unmatchable(op):
+                    continue
+                # didn't match explicit quantize handler class, we can check if the
+                # operator is in the related op set directly
+                if not _op_in_base_sets_of_related_ops(base_op):
```
@vkuzo please take a look at this change

vkuzo replied:
looks like this is now testing a smaller set of ops, could we keep the set of ops under test the same?  This would mean querying the structure you added (`get_native_backend_config_dict`) and taking the union of ops in there and ops in quantize handlers.

jerryzh168 replied:
not sure I follow, are you talking `get_default_quant_patterns` not returning the patterns in `get_native_backend_config_dict`?

vkuzo replied:
oh, my bad, I didn't realize that you added this in `get_all_quant_patterns`, sorry. Look good!

jerryzh168 replied:
oh I corrected it later after your comment, although some how the new patterns does appear as well, probably because we are modifying the patterns dictionary inplace..

jerryzh168 replied:
I'll change get_quant_patterns to return a copy, instead of the original dictionary

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXGraphMatcher

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXGraphMatcher

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXGraphMatcher

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

jerryzh168 committed:
Update on "[quant][fx] Support some default ops in the native backend config"


Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXGraphMatcher

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).
